From 2eb5c74a41495dccf87321ff70264321d6377f8b Mon Sep 17 00:00:00 2001
From: ValdikSS <iam@valdikss.org.ru>
Date: Sun, 9 Sep 2018 01:50:42 +0300
Subject: [PATCH] Increase maximum Bluetooth SBC codec bitpool and bitrate
 values

This commit increases maximum available bitpool value to 76 and maximum
possible bitrate to 455 kbit/s for 44.1 kHz, 496 kbit/s for 48 kHz, which is
optimal for both EDR 2 mbit/s (4 audio frames, 11.7 ms, 2 wasted bytes) and EDR
3 mbit/s (6 audio frames, 17.5 ms, 14 wasted bytes).

A2DP specification v1.2 states that all decoding devices (SNKs) should support
bitrates up to 512 kbit/s, and, in fact, some headphone manufacturers set
maximum bitpool values higher than 53.

Change-Id: I12557c312a5b720ece2ee0897997f392dd368802
---
 stack/a2dp/a2dp_sbc.cc         |  2 +-
 stack/a2dp/a2dp_sbc_encoder.cc | 16 ++++++++++++----
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/stack/a2dp/a2dp_sbc.cc b/stack/a2dp/a2dp_sbc.cc
index bef5bf961..f9a129a19 100644
--- a/stack/a2dp/a2dp_sbc.cc
+++ b/stack/a2dp/a2dp_sbc.cc
@@ -39,7 +39,7 @@
 #include "osi/include/log.h"
 #include "osi/include/osi.h"
 
-#define A2DP_SBC_MAX_BITPOOL 53
+#define A2DP_SBC_MAX_BITPOOL 76
 
 /* data type for the SBC Codec Information Element */
 typedef struct {
diff --git a/stack/a2dp/a2dp_sbc_encoder.cc b/stack/a2dp/a2dp_sbc_encoder.cc
index bd0998935..691e45b9c 100644
--- a/stack/a2dp/a2dp_sbc_encoder.cc
+++ b/stack/a2dp/a2dp_sbc_encoder.cc
@@ -38,8 +38,12 @@
 // A2DP SBC encoder interval in milliseconds.
 #define A2DP_SBC_ENCODER_INTERVAL_MS 20
 
-/* High quality quality setting @ 44.1 khz */
-#define A2DP_SBC_DEFAULT_BITRATE 328
+/*
+ * Higher quality setting. 495 kbps @ 48 khz, 454.8 kbps @ 44.1 khz.
+ * Up to 4 frames for 2DH5, 6 frames for 3DH5.
+ */
+#define A2DP_SBC_DEFAULT_BITRATE 455
+#define A2DP_SBC_48KHZ_BITRATE 496
 
 #define A2DP_SBC_NON_EDR_MAX_RATE 229
 
@@ -50,8 +54,8 @@
 #define MAX_2MBPS_AVDTP_MTU 663
 #define A2DP_SBC_MAX_PCM_ITER_NUM_PER_TICK 3
 
-#define A2DP_SBC_MAX_HQ_FRAME_SIZE_44_1 119
-#define A2DP_SBC_MAX_HQ_FRAME_SIZE_48 115
+#define A2DP_SBC_MAX_HQ_FRAME_SIZE_44_1 165
+#define A2DP_SBC_MAX_HQ_FRAME_SIZE_48 165
 
 /* Define the bitrate step when trying to match bitpool value */
 #define A2DP_SBC_BITRATE_STEP 5
@@ -835,6 +839,10 @@ static uint8_t calculate_max_frames_per_packet(void) {
 static uint16_t a2dp_sbc_source_rate() {
   uint16_t rate = A2DP_SBC_DEFAULT_BITRATE;
 
+  if (a2dp_sbc_encoder_cb.sbc_encoder_params.s16SamplingFreq == SBC_sf48000) {
+    rate = A2DP_SBC_48KHZ_BITRATE;
+  }
+
   /* restrict bitrate if a2dp link is non-edr */
   if (!a2dp_sbc_encoder_cb.is_peer_edr) {
     rate = A2DP_SBC_NON_EDR_MAX_RATE;
-- 
2.19.1

